### ISSUE TITLE
ZeRO-2 is broken due to a hook registration bug

### ISSUE THREAD

**Issue by AmericanPresidentJimmyCarter** [lodestone-rock/RamTorch#13]
# Bug: ZeRO-2 gradient reuse in `BouncingLinearFn.backward()` re-reduces stale owner-rank gradients

## Summary

When using `replace_linear_with_ramtorch()` together with ZeRO-2 gradient sharding (`setup_grad_sharding_hooks()`), gradient norms grow linearly with step count instead of remaining stable. Training effectively breaks within a few hundred steps.

The root cause is that `BouncingLinearFn.backward()` loads the previous `weight_cpu.grad` into the GPU accumulator **before** `_invoke_zero_2_tensor_hooks()` runs. That previous `.grad` has already been through ZeRO `dist.reduce(SUM)` on the owner rank. When it's mixed into the new raw gradient and passed through the ZeRO hook again, the old gradient contribution is re-reduced across ranks, causing the gradient magnitude to grow by approximately `world_size` per step.

## Affected code

`ramtorch/modules/linear.py`, `BouncingLinearFn.backward()`, lines 291-299:

```python
# Current (buggy) ordering:

# 1. Load previous CPU .grad onto GPU for accumulation (BEFORE ZeRO hooks)
w_grad_accum_buffers[selected_buffer] = (
    weight_cpu.grad.to(device, non_blocking=True)
    if weight_cpu.grad is not None
    else None
)

# ... later, the fresh gradient is computed and accumulated with this buffer ...

# 2. ZeRO hook reduces the combined (stale + fresh) gradient
w_grad_buffers[selected_buffer] = _invoke_zero_2_tensor_hooks(
    weight_cpu, w_grad_buffers[selected_buffer]
)

# 3. Write back to CPU
weight_cpu.grad = w_grad_buffers[selected_buffer].to("cpu", non_blocking=True)
```

The problem: step 1 loads a gradient that was already reduced by `dist.reduce(SUM)` on the previous step. Step 2 then reduces the sum of (stale reduced grad + fresh local grad) again, double-counting the old gradient.

## Reproduction

Minimal 2-GPU script — no dependencies beyond RAMTorch and PyTorch:

```python
#!/usr/bin/env python3
"""
torchrun --nproc_per_node=2 demonstrate_bug.py
"""
import os
import torch
import torch.distributed as dist
import torch.nn as nn
from ramtorch.helpers import replace_linear_with_ramtorch
from ramtorch.zero1 import create_zero_param_groups
from ramtorch.zero2 import setup_grad_sharding_hooks


def main():
    rank = int(os.environ["RANK"])
    world_size = int(os.environ["WORLD_SIZE"])
    dist.init_process_group(backend="nccl", rank=rank, world_size=world_size)
    torch.cuda.set_device(rank)
    device = torch.device(f"cuda:{rank}")

    torch.manual_seed(42)
    model = nn.Sequential(
        nn.Linear(256, 256), nn.ReLU(),
        nn.Linear(256, 256), nn.ReLU(),
        nn.Linear(256, 256), nn.ReLU(),
        nn.Linear(256, 256),
    )

    # Sync init weights
    model = model.to(device)
    for p in model.parameters():
        dist.broadcast(p.data, src=0)
    model = model.cpu()

    model = replace_linear_with_ramtorch(model, device=str(device))
    model = model.to(device)

    param_groups = [{"params": list(model.parameters()), "lr": 1e-3}]
    rank_param_groups = create_zero_param_groups(param_groups, world_size)
    setup_grad_sharding_hooks(rank_param_groups, rank)
    optimizer = torch.optim.SGD(rank_param_groups[rank], lr=1e-3)

    for step in range(20):
        torch.manual_seed(42 + step)
        x = torch.randn(8, 256, device=device)
        loss = model(x).pow(2).mean()
        loss.backward()
        torch.cuda.synchronize(device)

        # Measure global grad norm across ZeRO shards
        local_sq = sum(
            float(p.grad.float().pow(2).sum())
            for p in rank_param_groups[rank][0]["params"]
            if p.grad is not None
        )
        total = torch.tensor(local_sq, device=device)
        dist.all_reduce(total, op=dist.ReduceOp.SUM)
        norm = float(total.sqrt())

        if rank == 0:
            print(f"step {step:3d}  grad_norm = {norm:.4f}")

        optimizer.step()
        # NOT calling zero_grad — relying on RAMTorch's internal accumulation

        # Broadcast updated params
        with torch.no_grad():
            for owner, pgs in rank_param_groups.items():
                for g in pgs:
                    for p in g["params"]:
                        if p.device.type == "cuda":
                            dist.broadcast(p.data, src=owner)
                        else:
                            buf = p.data.to(device)
                            dist.broadcast(buf, src=owner)
                            p.data.copy_(buf.cpu())

    dist.destroy_process_group()


if __name__ == "__main__":
    main()
```

**Output (2 ranks):**

```
step   0  grad_norm = 0.0128
step   1  grad_norm = 0.0415
step   2  grad_norm = 0.0635
step   3  grad_norm = 0.0989
step   4  grad_norm = 0.1392
step   5  grad_norm = 0.1876
...
step  19  grad_norm = 1.6029
```

Gradient norm grows approximately linearly. With DDP using standard `nn.Linear`, the same setup stays flat at ~0.01.

**Adding `optimizer.zero_grad(set_to_none=True)` after `optimizer.step()` eliminates the growth** because it prevents the stale `.grad` from being loaded on the next backward. But this also disables gradient accumulation across microbatches, which is not always acceptable.

## Suggested fix

Move the `.grad` accumulation from **before** the ZeRO-2 hooks to **after** them, inside `BouncingLinearFn.backward()`:

```python
# In the "backward_grad_transfer" section (lines ~401-430):

with torch.cuda.stream(transfer_grad_stream):
    transfer_grad_stream.wait_event(compute_backward_finished_event)

    # 1. Run ZeRO-2 hooks on the FRESH gradient only
    w_grad_buffers[selected_buffer] = _invoke_zero_2_tensor_hooks(
        weight_cpu, w_grad_buffers[selected_buffer]
    )
    b_grad_buffers[selected_buffer] = _invoke_zero_2_tensor_hooks(
        bias_cpu, b_grad_buffers[selected_buffer]
    )

    # 2. THEN accumulate the previous .grad (which is already post-ZeRO)
    if w_grad_buffers[selected_buffer] is not None and weight_cpu.grad is not None:
        w_grad_buffers[selected_buffer] += weight_cpu.grad.to(device, non_blocking=True)
    if (
        bias_cpu is not None
        and b_grad_buffers[selected_buffer] is not None
        and bias_cpu.grad is not None
    ):
        b_grad_buffers[selected_buffer] += bias_cpu.grad.to(device, non_blocking=True)

    # 3. Write back to CPU
    if w_grad_buffers[selected_buffer] is not None:
        weight_cpu.grad = w_grad_buffers[selected_buffer].to("cpu", non_blocking=True)
    ...
```

And remove the pre-ZeRO accumulator loading (lines 291-299) that currently stages `.grad` into `w_grad_accum_buffers`/`b_grad_accum_buffers`.

This preserves the intended cross-microbatch gradient accumulation behavior without re-reducing stale owner-rank gradients.

## Also related: `broadcast_zero_params` skips RAMTorch parameters

`ramtorch/zero1.py`, `broadcast_zero_params()`, line 81:

```python
if not getattr(param, "is_ramtorch", False):
    work_handle = dist.broadcast(param.data, src=owner_rank, async_op=async_op)
```

This skips broadcasting all RAMTorch-backed parameters after `optimizer.step()`, which means ~95% of model parameters (all Linear weights/biases) are never synced across ranks. Each rank diverges after the first step.

The likely reason for the skip is that NCCL can't broadcast CPU tensors directly, but the fix is straightforward: stage CPU tensors through the local GPU device before broadcasting.

## Environment

- PyTorch 2.10
- RAMTorch (installed from git, latest as of 2026-03-16)
- CUDA 12.x, NCCL backend
- Tested on 2x and 5x NVIDIA 3090 GPUs

**Comment by AmericanPresidentJimmyCarter** [comment]
Followup: reproduction of the ZeRO-1 bug too.

```py
#!/usr/bin/env python3
"""
Minimal reproduction of the RAMTorch ZeRO-1 broadcast_zero_params bug.

Bug: broadcast_zero_params() skips all parameters with is_ramtorch=True.
After replace_linear_with_ramtorch(), that's every Linear weight and bias
(~95%+ of model params). So after optimizer.step(), each rank keeps its
own divergent copy — only the owner rank has the updated shard, other
ranks still have the stale pre-step value.

Effect on training: the model effectively becomes N different models
(one per rank), each seeing only 1/N of the gradient information and
never sharing updates. Loss appears flat because the averaged forward
outputs are incoherent.

Run with:
    torchrun --nproc_per_node=2 scripts/demonstrate_ramtorch_zero1_broadcast_bug.py
"""

from __future__ import annotations

import os

import torch
import torch.distributed as dist
import torch.nn as nn


def init():
    rank = int(os.environ["RANK"])
    world_size = int(os.environ["WORLD_SIZE"])
    dist.init_process_group(backend="nccl", rank=rank, world_size=world_size)
    torch.cuda.set_device(rank)
    return rank, world_size


def param_checksum(model) -> float:
    """Sum of all parameter values — diverges across ranks if broadcast is broken."""
    return sum(float(p.data.float().sum().item()) for p in model.parameters())


def main():
    rank, world_size = init()
    device = torch.device(f"cuda:{rank}")

    from ramtorch.helpers import replace_linear_with_ramtorch
    from ramtorch.zero1 import broadcast_zero_params, create_zero_param_groups
    from ramtorch.zero2 import setup_grad_sharding_hooks

    torch.manual_seed(42)
    model = nn.Sequential(
        nn.Linear(128, 128), nn.ReLU(),
        nn.Linear(128, 128), nn.ReLU(),
        nn.Linear(128, 128),
    )

    # Sync init weights across ranks
    model = model.to(device)
    for p in model.parameters():
        dist.broadcast(p.data, src=0)
    model = model.cpu()

    model = replace_linear_with_ramtorch(model, device=str(device))
    model = model.to(device)

    param_groups = [{"params": list(model.parameters()), "lr": 1e-2}]
    rank_param_groups = create_zero_param_groups(param_groups, world_size)
    setup_grad_sharding_hooks(rank_param_groups, rank)
    optimizer = torch.optim.SGD(rank_param_groups[rank], lr=1e-2)

    # Verify starting parity
    cksum_before = param_checksum(model)
    all_cksums = [None, None]
    dist.all_gather_object(all_cksums, cksum_before)
    if rank == 0:
        print(f"\n=== RAMTorch ZeRO-1 broadcast_zero_params Bug ===\n")
        print(f"Before training:")
        for r, c in enumerate(all_cksums):
            print(f"  rank {r} param checksum: {c:.6f}")
        match = "MATCH" if abs(all_cksums[0] - all_cksums[1]) < 1e-3 else "DIVERGED"
        print(f"  -> {match}\n")

    # Do a few training steps with upstream broadcast_zero_params
    for step in range(5):
        torch.manual_seed(42 + step)
        x = torch.randn(4, 128, device=device)
        loss = model(x).pow(2).mean()
        loss.backward()
        torch.cuda.synchronize(device)

        optimizer.step()
        optimizer.zero_grad(set_to_none=True)

        # Use upstream broadcast — this SKIPS is_ramtorch params
        broadcast_zero_params(rank_param_groups)

        cksum = param_checksum(model)
        all_cksums = [None, None]
        dist.all_gather_object(all_cksums, cksum)
        if rank == 0:
            delta = abs(all_cksums[0] - all_cksums[1])
            status = "MATCH" if delta < 1e-3 else "DIVERGED"
            print(f"Step {step} (upstream broadcast):  rank0={all_cksums[0]:+.4f}  rank1={all_cksums[1]:+.4f}  delta={delta:.4f}  {status}")

    if rank == 0:
        print()

    # Now show the fix: broadcast ALL params including CPU-backed ones
    # Reset to same init
    dist.barrier()
    torch.manual_seed(42)
    model2 = nn.Sequential(
        nn.Linear(128, 128), nn.ReLU(),
        nn.Linear(128, 128), nn.ReLU(),
        nn.Linear(128, 128),
    )
    model2 = model2.to(device)
    for p in model2.parameters():
        dist.broadcast(p.data, src=0)
    model2 = model2.cpu()
    model2 = replace_linear_with_ramtorch(model2, device=str(device))
    model2 = model2.to(device)

    param_groups2 = [{"params": list(model2.parameters()), "lr": 1e-2}]
    rank_param_groups2 = create_zero_param_groups(param_groups2, world_size)
    setup_grad_sharding_hooks(rank_param_groups2, rank)
    optimizer2 = torch.optim.SGD(rank_param_groups2[rank], lr=1e-2)

    def fixed_broadcast(rpg):
        """Broadcast ALL params, staging CPU tensors through GPU for NCCL."""
        with torch.no_grad():
            for owner_rank, pgs in rpg.items():
                for group in pgs:
                    for param in group["params"]:
                        if param.device.type == "cuda":
                            dist.broadcast(param.data, src=owner_rank)
                        else:
                            buf = param.data.to(device, non_blocking=False)
                            dist.broadcast(buf, src=owner_rank)
                            param.data.copy_(buf.to("cpu"), non_blocking=False)

    for step in range(5):
        torch.manual_seed(42 + step)
        x = torch.randn(4, 128, device=device)
        loss = model2(x).pow(2).mean()
        loss.backward()
        torch.cuda.synchronize(device)

        optimizer2.step()
        optimizer2.zero_grad(set_to_none=True)

        # Fixed broadcast — includes is_ramtorch params
        fixed_broadcast(rank_param_groups2)

        cksum = param_checksum(model2)
        all_cksums = [None, None]
        dist.all_gather_object(all_cksums, cksum)
        if rank == 0:
            delta = abs(all_cksums[0] - all_cksums[1])
            status = "MATCH" if delta < 1e-3 else "DIVERGED"
            print(f"Step {step} (fixed broadcast):    rank0={all_cksums[0]:+.4f}  rank1={all_cksums[1]:+.4f}  delta={delta:.4f}  {status}")

    if rank == 0:
        print()
        print("The upstream broadcast skips is_ramtorch params (line 81 of zero1.py):")
        print('    if not getattr(param, "is_ramtorch", False):')
        print()
        print("Effect: ranks diverge immediately. Each rank trains a different model.")
        print("Loss appears flat because averaged predictions are incoherent.")

    dist.destroy_process_group()


if __name__ == "__main__":
    main()
```

**Comment by AmericanPresidentJimmyCarter** [comment]
I benchmarked the difference between the broken upstream and my working local branch too:

```
Benchmarking RAMTorch ZeRO overhead: 5 GPUs
Model: ~18.9M params

                       Mean (ms)   Median (ms)    Min (ms)    Max (ms)
----------------------------------------------------------------------
   Upstream (broken)      503.12        500.27      304.60      736.07
       Local (fixed)      769.46        771.45      450.11     1138.69

Overhead (mean):   +52.9%
Overhead (median): +54.2%

Note: upstream is 'faster' only because it skips broadcasting ~95% of params.
That speed comes at the cost of completely broken training.
```

**Comment by lodestone-rock** [comment]
im trying to create a better version of this.
currently working with torch run **WILL ALWAYS HAS THIS BUG**

because the model is not sharing the same memory space in the CPU. the only way to use this ramtorch is to call torch.multiprocess and spawn the process **IN THE CODE**

but if you use torchrun then it will spawn an individual process where it has its own memory space and this will break the CPU weight sharing 

ideally we should not use torch run because it complicates the code and gradient orchestration 
so within node the training should be a single process multiple thread style training instead of spawning many child process.

**Comment by lodestone-rock** [comment]
but this means breaking the classical flow of torchrun style code, which is not ideal.
but i guess it's better to have distinct code style when using ramtorch instead of trying to fit square peg into round hole.

**Comment by lodestone-rock** [comment]
[https://github.com/lodestone-rock/x0-pred/blob/main/celeba_ramtorch.py#L706-L710](https://github.com/lodestone-rock/x0-pred/blob/main/celeba_ramtorch.py#L706-L710)

you need to init the model before spawning the process
that way it will share the same memory space 

since we must do this in order to use the same address space then it's better to just create better flow where we use thread instead of multiprocess style training.
